src/tags/object/Video/TransformTools.ts

```diff
@@ -0,0 +1,116 @@
+import { KonvaEventObject } from "konva/lib/Node";
+import { Box, Transformer } from "konva/lib/shapes/Transformer";
+import { Stage } from "konva/lib/Stage";
+import { WorkingArea } from "./types";
+
+// define several math function
+export const getCorner = (pivotX: number, pivotY: number, diffX: number, diffY: number, angle: number) => {
+  const distance = Math.sqrt(diffX * diffX + diffY * diffY);
+
+  /// find angle from pivot to corner
+  angle += Math.atan2(diffY, diffX);
+
+  /// get new x and y and round it off to integer
+  const x = pivotX + distance * Math.cos(angle);
+  const y = pivotY + distance * Math.sin(angle);
+
+  return { x, y };
+};
+
+export const getClientRect = (rotatedBox: Box) => {
+  const { x, y, width, height } = rotatedBox;
+  const rad = rotatedBox.rotation;
+
+  const p1 = getCorner(x, y, 0, 0, rad);
+  const p2 = getCorner(x, y, width, 0, rad);
+  const p3 = getCorner(x, y, width, height, rad);
+  const p4 = getCorner(x, y, 0, height, rad);
+
+  const minX = Math.min(p1.x, p2.x, p3.x, p4.x);
+  const minY = Math.min(p1.y, p2.y, p3.y, p4.y);
+  const maxX = Math.max(p1.x, p2.x, p3.x, p4.x);
+  const maxY = Math.max(p1.y, p2.y, p3.y, p4.y);
+
+  return {
+    x: minX,
+    y: minY,
+    width: maxX - minX,
+    height: maxY - minY,
+  };
+};
+
+export const getTotalBox = (boxes: {
+  x: number,
+  y: number,
+  width: number,
+  height: number,
+}[]) => {
+  let minX = Infinity;
+  let minY = Infinity;
+  let maxX = -Infinity;
+  let maxY = -Infinity;
+
+  boxes.forEach((box) => {
+    minX = Math.min(minX, box.x);
+    minY = Math.min(minY, box.y);
+    maxX = Math.max(maxX, box.x + box.width);
+    maxY = Math.max(maxY, box.y + box.height);
+  });
+
+  return {
+    x: minX,
+    y: minY,
+    width: maxX - minX,
+    height: maxY - minY,
+  };
+};
+
+export const boundBoxFunction = (workingArea: WorkingArea, enabled = true) => (oldBox: Box, newBox: Box) => {
+  if (!enabled) return newBox;
+
+  const box = getClientRect(newBox);
+  const isOut =
+    (box.x - workingArea.x) < 0 ||
+    (box.y - workingArea.y) < 0 ||
+    (box.x - workingArea.x) + box.width > workingArea.width ||
+    (box.y - workingArea.y) + box.height > workingArea.height;
+
+  // if new bounding box is out of visible viewport, let's just skip transforming
+  // this logic can be improved by still allow some transforming if we have small available space
+  if (isOut) {
+    return oldBox;
+  }
+  return newBox;
+};
+
+export const onDragMove = (workingArea: WorkingArea, enabled = true) => function(this: Transformer, e: KonvaEventObject<DragEvent>) {
```

Comment:
it returns function, so name should reflect this — like `createOnDragMoveHandler` or maybe something shorter